lisa/tests/staging/sched_android.py

```diff
@@ -0,0 +1,239 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import os.path
+import abc
+
+from lisa.wlgen.rta import RTA, Periodic
+from lisa.tests.base import TestBundle, Result, ResultBundle, RTATestBundle
+from lisa.trace import Trace, FtraceCollector, FtraceConf, requires_events
+from lisa.target import Target
+from lisa.utils import ArtifactPath
+
+class SchedTuneItemBase(RTATestBundle):
+    """
+    Abstract class enabling rtapp execution in a schedtune group
+
+    :param boost: The boost level to set for the cgroup
+    :type boost: int
+
+    :param prefer_idle: The prefer_idle flag to set for the cgroup
+    :type prefer_idle: bool
+    """
+
+    def __init__(self, res_dir, plat_info, boost, prefer_idle):
+        super().__init__(res_dir, plat_info)
+        self.boost = boost
+        self.prefer_idle = prefer_idle
+
+    @property
+    def cgroup_configuration(self):
+        return self.get_cgroup_configuration(self.plat_info, self.boost, self.prefer_idle)
+
+    @classmethod
+    def get_cgroup_configuration(cls, plat_info, boost, prefer_idle):
+        values = {
+                    'boost': boost,
+                    'prefer_idle': int(prefer_idle)
+                 }
+        return { 'name': 'lisa_test',
+                 'controller': 'schedtune',
+                 'values': values }
+
+    @classmethod
+    # Not annotated, to prevent exekall from picking it up. See
+    # SchedTuneBase.from_target
+    def from_target(cls, target, boost, prefer_idle, res_dir=None, ftrace_coll=None):
+        """
+        .. warning:: `res_dir` is at the end of the parameter list, unlike most
+            other `from_target` where it is the second one.
+        """
+        return super().from_target(target, res_dir, boost=boost,
+                prefer_idle=prefer_idle, ftrace_coll=ftrace_coll)
+
+    @classmethod
+    def _from_target(cls, target, res_dir, boost, prefer_idle, ftrace_coll=None):
+        plat_info = target.plat_info
+        rtapp_profile = cls.get_rtapp_profile(plat_info)
+        cgroup_config = cls.get_cgroup_configuration(plat_info, boost, prefer_idle)
+        cls._run_rtapp(target, res_dir, rtapp_profile, ftrace_coll, cgroup_config)
+
+        return cls(res_dir, plat_info, boost, prefer_idle)
+
+class SchedTuneBase(TestBundle):
+    """
+    Abstract class enabling the aggregation of ``SchedTuneItemBase``
+
+    :param test_bundles: a list of test bundles generated by
+        multiple ``SchedTuneItemBase`` instances
+    :type test_bundles: list
+    """
+    def __init__(self, res_dir, plat_info, test_bundles):
+        super().__init__(res_dir, plat_info)
+
+        self.test_bundles = test_bundles
+
+    @classmethod
+    def from_target(cls, target:Target, res_dir:ArtifactPath=None,
+            ftrace_coll:FtraceCollector=None) -> 'SchedTuneBase':
+        """
+        XXX: Warning if nothing is documented here ...
```

Comment:
I don't understand why I need to have something there. With no docstring there is a 'Warning that must be fixed' with `pedantic_build.sh` ...

Reply:
I think I got it - the warning is `WARNING: py:meth reference target not found: _from_target`

If you don't define a docstring, you inherit one, and it happens that the one you inherent from `TestBundle` uses 
```
:meth:`_from_target`
```
instead of
```
:meth:`lisa.tests.base.TestBundle._from_target`
```

so it tries to find a local definition of `lisa.tests.staging.sched_android.SchedTuneBase._from_target`. Using full paths in the docstring of `TestBundle` should fix it (but then the doc would link to `TestBundle._from_target` even if you redefine the docstring of `_from_target` in your own class.... Bah).